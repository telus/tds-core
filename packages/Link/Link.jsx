import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { colorGreyShark, colorWhite, colorGainsboro, colorGreyRaven } from '@tds/core-colours'
import { links } from '@tds/shared-styles'
import { safeRest } from '@tds/util-helpers'
import { componentWithName } from '@tds/util-prop-types'

import { withForwardedRef } from '@tds/shared-hocs'

import { warn } from '../../shared/utils/warn'

const base = {
  ...links.focusOutline,
  '&:link,&:visited': {
    color: colorGreyShark,
    textDecoration: 'underline',
  },
  '&:hover': {
    textDecoration: 'none',
  },
  '& svg': {},
}

const states = ({ invert }) => {
  return {
    '&:active': {
      color: invert && colorGainsboro,
      backgroundColor: invert ? 'rgba(0,0,0,0.4)' : colorGainsboro,
      borderRadius: '0.25rem',
      padding: '0.125rem',
      margin: '-0.125rem',
      textDecoration: 'underline',
    },
    '&:focus': {
      border: `0.125rem solid ${invert ? colorWhite : colorGreyRaven}`,
      padding: '0.125rem',
      margin: '-0.25rem', // (border + padding) * -1
      borderRadius: '0.25rem',
      outline: 'none',
    },
  }
}
const StyledLink = styled.a(
  base,
  {
    '& svg': {
      transition: 'transform 150ms ease-in-out',
    },
    '&:hover svg': {
      transform: 'scale(1.1, 1.1)',
    },
    '&:active svg': {
      transform: 'scale(1, 1)',
    },
  },
  ({ invert, context }) => {
    if (context.inheritColor) {
      return {
        '&:link,&:visited': {
          color: 'inherit',
        },
      }
    }
    if (invert) {
      return {
        '&:link,&:visited': {
          color: colorWhite,
        },
      }
    }
    return {}
  },
  states,
  ({ hasIcon }) => {
    if (hasIcon) {
      return {
        display: 'inline-block',
        '& > svg': {
          verticalAlign: 'bottom',
        },
      }
    }
    return {}
  }
)

/**
 * @version ./package.json
 */
const Link = (
  { reactRouterLinkComponent, invert, children, forwardedRef, icon: Icon, iconPosition, ...rest },
  context
) => {
  if (!(reactRouterLinkComponent && rest.to) && (reactRouterLinkComponent || rest.to)) {
    warn('Link', 'The props `reactRouterLinkComponent` and `to` must be used together.')
  }

  const renderChildren = useCallback(() => {
    if (Icon) {
      return (
        <>
          {iconPosition === 'left' && (
            <Icon color={invert ? 'white' : 'greyShark'} style={{ marginRight: '0.5rem' }} />
          )}
          {children}
          {iconPosition === 'right' && (
            <Icon color={invert ? 'white' : 'greyShark'} style={{ marginLeft: '0.25rem' }} />
          )}
        </>
      )
    }
    return children
  }, [children, Icon, iconPosition, invert])

  return (
    <StyledLink
      {...safeRest(rest)}
      as={reactRouterLinkComponent || 'a'}
      invert={invert}
      context={context}
      ref={forwardedRef}
      hasIcon={!!Icon}
    >
      {renderChildren()}
    </StyledLink>
  )
}
Link.propTypes = {
  /**
   * React Router Link component.
   */
  reactRouterLinkComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  /**
   * Target URL (if using 'reactRouterLinkComponent')
   */
  to: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  /**
   * Target URL.
   */
  href: PropTypes.string,
  /**
   * Invert link style to appear light on dark backgrounds.
   */
  invert: PropTypes.bool,
  /**
   * Link text.
   */
  children: PropTypes.node.isRequired,
  /**
   * @ignore
   */
  forwardedRef: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
  /**
   * Provide an icon from the Dependent icon group in `@tds/core-interactive-icon`.
   * @since 2.2.0
   */
  icon: componentWithName('Dependent', true),
  /**
   * When `icon` is provided, use `iconPosition` to place the Icon to the left or right side of Link.
   * @since 2.2.0
   */
  iconPosition: PropTypes.oneOf(['left', 'right']),
}
Link.defaultProps = {
  reactRouterLinkComponent: null,
  to: null,
  href: null,
  invert: undefined,
  forwardedRef: undefined,
  icon: undefined,
  iconPosition: 'left',
}

Link.contextTypes = {
  inheritColor: PropTypes.bool,
}

export default withForwardedRef(Link)
