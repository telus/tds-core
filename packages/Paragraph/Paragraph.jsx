import React from 'react'
import PropTypes from 'prop-types'

import typographyStyles from '@tds/shared-typography/Typography.modules.scss'

import safeRest from '../../shared/utils/safeRest'
import joinClassNames from '../../shared/utils/joinClassNames'

import spacingStyles from '../../shared/styles/Spacing.modules.scss'
import styles from './Paragraph.modules.scss'

/**
 * Block text as an HTML `<p>` element.
 *
 * @version ./package.json
 */
const Paragraph = ({ bold, size, align, invert, children, ...rest }, context) => {
  const paragraphColor = invert ? typographyStyles.invertedColor : typographyStyles.color

  const classes = joinClassNames(
    typographyStyles.wordBreak,
    spacingStyles.noSpacing,
    context.inheritColor ? typographyStyles.inheritColor : paragraphColor,
    typographyStyles[size],
    bold ? typographyStyles.boldFont : typographyStyles[`${size}Font`],
    styles[`${align}Align`],
    typographyStyles.default
  )

  return (
    <p {...safeRest(rest)} className={classes}>
      {children}
    </p>
  )
}

Paragraph.propTypes = {
  /**
   * Embolden paragraph text without conveying any special importance or relevance.
   */
  bold: PropTypes.bool,
  /**
   * The font size.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Content alignment.
   */
  align: PropTypes.oneOf(['left', 'center', 'right']),
  /**
   * Invert the text color to appear light on dark backgrounds.
   */
  invert: PropTypes.bool,
  /**
   * The content. Can be text, other components, or HTML elements.
   */
  children: PropTypes.node.isRequired,
}

Paragraph.defaultProps = {
  bold: false,
  size: 'medium',
  align: 'left',
  invert: false,
}

Paragraph.contextTypes = {
  inheritColor: PropTypes.bool,
}

export default Paragraph
