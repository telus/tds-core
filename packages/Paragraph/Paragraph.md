```
<Box between={3}>
  <Paragraph>
    Our commitment to service is demonstrated in everything we do.
  </Paragraph>
  <Paragraph>
    Enjoy your choice of <Link href="#">5 channels</Link>, and up to <Link href="#">23 local and regional channels</Link>. Plus, add Premium channel packs like HBO, TMN &amp; Crave TV for just $20/mo. All delivered via TELUS Internet through the Pik TV media box and Pik TV app.<Text.Sup>2</Text.Sup>
  </Paragraph>
</Box>
```

### Placing text on dark colours

Use the `invert` prop.

```jsx { "props": { "className": "docs_purple-block" } }
<Paragraph invert>
  Order online and save an extra $150.
</Paragraph>
```
