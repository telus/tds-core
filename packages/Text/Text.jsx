import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../shared/utils/safeRest'
import joinClassNames from '../../shared/utils/joinClassNames'

import TextSup from './TextSup/TextSup'

import styles from '../../shared/styles/Typography/Typography.modules.scss'

/**
 * Inline text usually as an HTML `<span>` element.
 *
 * @version ./package.json
 */
const Text = ({ block, bold, size, invert, children, ...rest }, context) => {
  const textColor = invert ? styles.invertedColor : styles.color

  const classes = joinClassNames(
    styles[size],
    bold ? styles.boldFont : styles[`${size}Font`],
    context.inheritColor ? styles.inheritColor : textColor,
    block ? styles.blockText : null
  )

  return (
    <span {...safeRest(rest)} className={classes}>
      {children}
    </span>
  )
}

Text.propTypes = {
  /**
   * If true, renders a block level element.
   * Otherwise, renders an inline element.
   */
  block: PropTypes.bool,
  /**
   * Embolden text without conveying any special importance or relevance.
   */
  bold: PropTypes.bool,
  /**
   * The font size.
   */
  size: PropTypes.oneOf(['base', 'small', 'medium', 'large']),
  /**
   * Invert the text color to appear light on dark backgrounds.
   */
  invert: PropTypes.bool,
  /**
   * The content. Can be text, other components, or HTML elements.
   */
  children: PropTypes.node.isRequired,
}

Text.defaultProps = {
  block: false,
  bold: false,
  size: 'base',
  invert: false,
}

Text.contextTypes = {
  inheritColor: PropTypes.bool,
}

Text.Sup = TextSup

export default Text
