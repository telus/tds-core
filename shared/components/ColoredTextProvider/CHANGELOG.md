# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 1.1.3 (2022-02-16)


### Bug Fixes

* gha release job npm publish token ([ce7aacd](https://github.com/telus/tds-core-test/commit/ce7aacdd77e54c4f8e48eb31d0c05dc804530324))





## 1.1.2 (2022-02-15)


### Bug Fixes

* gha release job git config ([276ebb6](https://github.com/telus/tds-core-test/commit/276ebb6968a0b56c9b87c178f6895a60ae108e71))





## 1.1.1 (2022-02-15)


### Bug Fixes

* gha release job git config ([1ab603d](https://github.com/telus/tds-core-test/commit/1ab603d68c36219b0711fc353bc2515b64712ca9))





# [1.1.0](https://github.com/telus/tds-core/compare/@tds/shared-colored-text-provider@1.0.1...@tds/shared-colored-text-provider@1.1.0) (2019-04-29)


### Features

* **shared-colored-text-provider:** enable className prop ([62bba23](https://github.com/telus/tds-core/commit/62bba23))





## [1.0.1](https://github.com/telus/tds-core/compare/@tds/shared-colored-text-provider@1.0.0...@tds/shared-colored-text-provider@1.0.1) (2019-02-06)

**Note:** Version bump only for package @tds/shared-colored-text-provider





<a name="1.0.0"></a>
# 1.0.0 (2018-12-06)


### Features

* **shared:** add tag property to ColoredTextProvider ([e77b659](https://github.com/telus/tds-core/commit/e77b659))
* **shared-colored-text-provider:** create new private package ([813b6fa](https://github.com/telus/tds-core/commit/813b6fa))
