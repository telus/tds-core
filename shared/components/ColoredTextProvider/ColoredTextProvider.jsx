import React from 'react'
import PropTypes from 'prop-types'

class ColoredTextProvider extends React.Component {
  getChildContext() {
    return {
      inheritColor: true,
    }
  }

  render() {
    const { colorClassName, tag, children } = this.props

    return React.createElement(tag, { className: colorClassName }, children)
  }
}

ColoredTextProvider.propTypes = {
  colorClassName: PropTypes.string.isRequired,
  tag: PropTypes.string,
  children: PropTypes.node.isRequired,
}

ColoredTextProvider.defaultProps = {
  tag: 'div',
}

ColoredTextProvider.childContextTypes = {
  inheritColor: PropTypes.bool,
}

export default ColoredTextProvider
