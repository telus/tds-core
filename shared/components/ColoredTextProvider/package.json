{
  "name": "@tds/shared-colored-text-provider",
  "version": "1.0.1",
  "private": true,
  "main": "",
  "module": ""
}
