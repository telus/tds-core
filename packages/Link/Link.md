### Minimal usage

```jsx
<Link href="#">Go to the TDS homepage</Link>
```

### Usage criteria

- Links inherit the font size of surrounding text.
- Their colour is either shuttle grey or white.

### Inverted

```jsx { "props": { "className": "docs_purple-block" } }
<Paragraph invert>
  Call Screen requires a subscription.{' '}
  <Link href="#" invert>
    Log in
  </Link>{' '}
  to your account and make sure you are subscribed.
</Paragraph>
```

### Using with React Router Links

When using Link with React Router, you must pass in the React Router Link component as a prop. Our intention is to have developers supply the React Router Link component whether they are on version 3 of React Router, or are using the latest React Router DOM package on version 4. This way, we avoid deep dependency conflicts.

It is recommended to set up a component wrapper in your project to avoid repetition, like so:

**LinkWrapper.jsx**

```jsx noeditor static
import React from 'react'

// React Router version 3
import { Link as ReactRouterLink } from 'react-router'

// React Router version 4
// import { Link as ReactRouterLink } from 'react-router-dom'

import TdsLink from '@tds/core-link'

// If you're importing from @telus/tds, write this as:
// import { Link as TdsLink } from '@telus/tds';

const LinkWrapper = ({ children, ...rest }) => (
  <TdsLink {...rest} reactRouterLinkComponent={rest.to ? ReactRouterLink : undefined}>
    {children}
  </TdsLink>
)

export default LinkWrapper
```

**App.js**

```js noeditor static
import Link from './LinkWrapper'

const MyApp = () => (
  <main>
    ...
    <Link to="/more">Read More</Link>
    <Link href="https://www.telus.com/">Telus Website</Link>
  </main>
)
```

This way, you can output normal anchors or react router links using a single component.
