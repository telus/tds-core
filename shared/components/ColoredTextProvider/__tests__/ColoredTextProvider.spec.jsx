/* eslint no-console: "error" */
import React from 'react'
import { mount } from 'enzyme'

import Paragraph from '../../../../packages/Paragraph/Paragraph'
import Link from '../../../../packages/Link/Link'
import Text from '../../../../packages/Text/Text'
import ColoredTextProvider from '../ColoredTextProvider'

describe('ColoredTextProvider', () => {
  it('allows Paragraph to inherit its color', () => {
    const coloredText = mount(
      <ColoredTextProvider colorClassName="some-class-name">
        <Paragraph>Some content</Paragraph>
      </ColoredTextProvider>
    )

    expect(coloredText.find('p')).toHaveClassName('inheritColor')
  })

  it('allows Link to inherit its color', () => {
    const coloredText = mount(
      <ColoredTextProvider colorClassName="some-class-name">
        <Link href="http://fake.com">a hyperlink</Link>
      </ColoredTextProvider>
    )
    expect(coloredText.find('a')).toMatchSnapshot()
  })

  it('allows Text to inherit its color', () => {
    const coloredText = mount(
      <ColoredTextProvider colorClassName="some-class-name">
        <Text>Some inline content</Text>
      </ColoredTextProvider>
    )

    expect(coloredText.find('span')).toHaveClassName('inheritColor')
  })
})
