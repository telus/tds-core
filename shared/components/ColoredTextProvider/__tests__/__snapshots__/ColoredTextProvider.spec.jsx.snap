// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColoredTextProvider allows Link to inherit its color 1`] = `
.c0:focus {
  outline: auto 5px -webkit-focus-ring-color;
}

.c0:link,
.c0:visited {
  color: #2a2c2e;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:link,
.c0:visited {
  color: inherit;
}

<a
  className="c0"
  href="http://fake.com"
  to={null}
>
  a hyperlink
</a>
`;
